components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesBookmarksStorage.kt

```diff
@@ -0,0 +1,177 @@
+     *
+     * @param root The bookmark root to obtain.
+     * @param recursive Whether to recurse and obtain all levels of children.
+     * @return The populated root.
+     */
+    override suspend fun getTree(root: BookmarkRoot, recursive: Boolean): BookmarkTreeNode? {
+        return withContext(scope.coroutineContext) {
+            reader.getBookmarksTree(root.id, recursive)
+        }
+    }
+
+    /**
+     * Produces a populated bookmarks tree for the given guid string.
+     *
+     * @param guid The bookmark guid to obtain.
+     * @param recursive Whether to recurse and obtain all levels of children.
+     * @return The populated root starting from the guid.
+     */
+    override suspend fun getTree(guid: String, recursive: Boolean): BookmarkTreeNode? {
+        return withContext(scope.coroutineContext) {
+            reader.getBookmarksTree(guid, recursive)
+        }
+    }
+
+    /**
+     * Searches bookmarks with a query string.
+     *
+     * @param query The query string to search.
+     * @param limit The maximum number of entries to return.
+     * @return The list of matching BookmarkItems.
+     */
+    override suspend fun searchBookmarks(query: String, limit: Int): List<BookmarkItem> {
+        return withContext(scope.coroutineContext) {
+            reader.searchBookmarks(query, limit)
+        }
+    }
+
+    /**
+     * Adds a new bookmark tree node to a given node.
+     *
+     * @param newNode The node to add.
+     * @param position The optional position to add it in.
+     * @return The guid of the newly inserted bookmark item.
+     */
+    override suspend fun addNode(target: BookmarkFolder, newNode: BookmarkTreeNode, position: Int?): String {
+        return withContext(scope.coroutineContext) {
+            when (newNode) {
+                is BookmarkFolder -> {
+                    writer.createFolder(target.guid, newNode.title ?: "", position)
+                }
+                is BookmarkItem -> {
+                    writer.createBookmarkItem(target.guid, newNode.url, newNode.title, position)
+                }
+                is BookmarkSeparator -> {
+                    writer.createSeparator(target.guid, position)
+                }
+                else -> throw IllegalStateException("Unsupported bookmark node type for addNode")
+            }
+        }
+    }
+
+    /**
+     * Moves a bookmark tree node to a given node.
+     *
+     * @param newFolder The node to move it to.
+     * @param position The optional position to move it to.
+     * @return The guid of the moved node.
+     */
+    override suspend fun moveNode(nodeToMove: BookmarkTreeNode, newFolder: BookmarkFolder, position: Int?): String {
+        return withContext(scope.coroutineContext) {
+            when (nodeToMove) {
+                is BookmarkFolder -> {
+                    val title = nodeToMove.title
+                    writer.deleteBookmarkNode(nodeToMove.guid)
```

Comment:
None of these API calls are correct.

You should be instead using `updateBookmark`, and constructing a `BookmarkUpdateInfo` which describes what a move looks like.

Reply:
I missed that part implied in the docs that updateBookmark also performed moves. Thanks for catching this.